test/decl/typealias/generic.swift

```diff
@@ -89,7 +89,7 @@ let _ : D<Int, Int, Float> = D(a: 1, b: 2)
 let _ : F = { (a : Int) -> Int in a }  // Infer the types of F
 
 // TODO QoI: Cannot infer T1/T2.
-let _ : F = { a in a }  // expected-error {{cannot convert value of type '(_) -> _' to specified type 'F'}}
+let _ : F = { a in a }  // expected-error {{type of expression is ambiguous without more context}}
```

Comment:
There is some logic there to walk the caller and callee argument types in parallel, matching up archetypes and collecting mismatches. See diagnoseArchetypeAmbiguity(), diagnoseUnboundArchetype(), findGenericSubstitutions(), and diagnoseGenericParameterErrors(). I don't understand how they all fit together and it all looks very unprincipled. I suspect we could get better diagnostics with 1/3rd of the code if we just cleaned this up.